internal/server.go

```diff
@@ -11,50 +11,142 @@ import (
-user information.
-
-
-## Abuse
-
-If, for whatever reason, you would like to have a file removed prematurely,
-please write an e-mail to <{{.EMail}}>.
-
-Please allow me a certain amount of time to react and work on your request.
-`
+const indexTpl = `<html>
+<head>
+  <style>
+    * {
+      font-family: monospace;
+    }
+
+    body {
+      margin: 0 auto;
+      padding: 1rem;
+      width: 50%;
+    }
+
+    h1 {
+      padding-top: 3rem;
+    }
+
+    h2 {
+      padding-top: 2rem;
+    }
+
+    h3 {
+      padding-top: 1rem;
+    }
+
+    pre {
+      background-color: #eee;
+      padding: 0.5rem;
+    }
+
+    form {
+      padding: 0.5rem;
+      position: relative;
+      margin: auto;
+      box-sizing: border-box;
+      background-color: #eee;
+    }
+
+    #grid {
+      display: grid;
+      grid-gap: 1rem;
+      grid-template-columns: 1fr 1fr;
+      grid-template-rows: repeat(3, 2rem);
+      margin-bottom: 1rem;
+    }
+
+    #grid > * {
+      margin: auto 0;
+    }
+
+    input, button{
+      width: 100%;
+    }
+  </style>
+</head>
+
+<body>
+  <h1># gosh! Go Share</h1>
+  <p>
+    Upload your files to this server and share them with your friends or, if
+    non-existent, shady people from the Internet.
+  </p>
+  <p>
+    Your file will expire after
+    {{.Expires}} or earlier, if explicitly specified. Optionally, the file can
+    be deleted directly after the first retrieval. In addition, the maximum file
+    size is {{.Size}}. 
+  </p>
+  <p>
+    This is no place to share questionable or illegal data.
+    Please use another service or stop it completely. Get some help.
+  </p>
+  <p>
+    The gosh
+    software can be obtained from
+    <a href="https://github.com/oxzi/gosh">https://github.com/oxzi/gosh</a>
+  </p>
+
+  <h2>## Posting</h2>
+
+  <h3>### curl</h3>
+
+  HTTP POST your file:
+
+  <pre>$ curl -F 'file=@foo.png' {{.Proto}}://{{.Hostname}}/</pre>
+
+  Burn after reading:
+
+  <pre>$ curl -F 'file=@foo.png' -F 'burn=1' {{.Proto}}://{{.Hostname}}/</pre>
+
+  Set a custom expiry date, e.g., one minute:
+
+  <pre>
+$ curl -F 'file=@foo.png' -F 'time=1m' {{.Proto}}://{{.Hostname}}/</pre>
+
+  Or all together:
+
+  <pre>
+$ curl -F 'file=@foo.png' -F 'time=1m' -F 'burn=1' {{.Proto}}://{{.Hostname}}/</pre>
+
+  <h3>### form</h3>
+  
+  <form action="{{.Proto}}://{{.Hostname}}/" method="POST" enctype="multipart/form-data">
+    <div id="grid">
+      <label for="file">Your file:</label>
+      <input type="file" name="file" />
+      <label for="burn">Burn after reading:</label>
+      <input type="text" name="burn" pattern="1?" title="Use '1' to burn, leave empty to retain your file" />
+      <label for="time">Optionally, set a custom expiry date:</label>
+      <input
+        type="text"
+        name="time"
+        pattern="[-+]?([0-9]+(\.[0-9]*)?(ns|us|ms|s|m|h)+)+"
+        title="A duration string is a possibly signed sequence of decimal numbers,
```

Comment:
The server does not parse a time.Duration, only a variant.

https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L63-L67

This does not allow negative durations, but days, weeks, months or years. Thus, you might want to adjust both the pattern and title to the one in this method's documentation.

Reply:
Addressed in the updated review.

Reply:
Q: So what would be valid time strings? Is "1mo 3d" valid, or only single dateinterval terms like "1mo" or "1d" on their own?

Reply:
Any ordered sequence of duration pairs defined by a positive decimal number and a time unit specifier, e.g., `1h30m` or `23m42s`.

All possible time unit specifiers are defined in `internal/util.go`.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L28-L38

Furthermore, some examples are available in the test. Please notice third `valid` field which indicates if one testcase represents a valid duration.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util_test.go#L19-L29

Finally, the `getDurationPattern` function generates a regular expression for such a string.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L45-L61